### PR TITLE
Uses coroutines in ProductHunt. Introduces a remote data source for P…

### PR DESCRIPTION
**Pull request by florina-muntenescu**
…roduct Hunt

### CODE REVIEW

File: core/src/main/java/io/plaidapp/core/producthunt/data/api/ProductHuntRepository.kt [hunk 1]
```diff
@@ -16,53 +16,72 @@
...
+ * Class that knows how to get Product Hunt posts
+ */
+class ProductHuntRepository(
+    private val remoteDataSource: ProductHuntRemoteDataSource,
+    private val dispatcherProvider: CoroutinesDispatcherProvider
+) {
+    private val parentJob = Job()
+    private val scope = CoroutineScope(dispatcherProvider.main + parentJob)
+
+    private val inflight = mutableMapOf<Int, Job>()
```
Comment (nickbutcher) [inline core/src/main/java/io/plaidapp/core/producthunt/data/api/ProductHuntRepository.kt]: Can't we remove this now that we have a scope?

File: core/src/test/java/io/plaidapp/core/producthunt/data/ProductHuntRemoteDataSourceTest.kt [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import retrofit2.Response
+
+/**
+ * Tests for [ProductHuntRemoteDataSource] that mock the Product Hunt API
+ */
+class ProductHuntRemoteDataSourceTest {
+
+    private val errorResponseBody = ResponseBody.create(MediaType.parse(""), "Error")
+
+    private val post1 = Post(
```
Comment (nickbutcher) [inline core/src/test/java/io/plaidapp/core/producthunt/data/ProductHuntRemoteDataSourceTest.kt]: Named args for readability?

File: core/src/test/java/io/plaidapp/core/producthunt/data/ProductHuntRemoteDataSourceTest.kt [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    fun loadData_whenResultSuccessful() = runBlocking {
+        // Given that the service responds with success
+        val result = Response.success(postsResult)
+        whenever(service.getPosts(1)).thenReturn(CompletableDeferred(result))
+
+        // When loading the data
+        val response = dataSource.loadData(1)
+
+        // Then the response is the expected one
+        Assert.assertNotNull(response)
```
Comment (nickbutcher) [inline core/src/test/java/io/plaidapp/core/producthunt/data/ProductHuntRemoteDataSourceTest.kt]: Statically import method

**Comment by florina-muntenescu** [comment]
@nickbutcher & @ataulm, ready for another 👀

File: core/src/main/java/io/plaidapp/core/producthunt/data/api/ProductHuntRepository.kt [hunk 4]
```diff
@@ -16,53 +16,68 @@
...
+            when (result) {
+                is Result.Success -> onSuccess(result.data)
+                is Result.Error -> onError(result.exception.toString())
+            }.exhaustive
+        }
     }
 
     fun cancelAllRequests() {
-        for (request in inflight.values) request.cancel()
+        parentJob.cancelChildren()
```
Comment (ataulm) [inline core/src/main/java/io/plaidapp/core/producthunt/data/api/ProductHuntRepository.kt]: :+1:

**Pull request merged**